lib/diaspora_federation/entities/relayable.rb

```diff
@@ -4,10 +4,56 @@ module Relayable
       def self.included(model)
         model.class_eval do
           property :parent_guid
-          property :parent_author_signature
-          property :author_signature
+          property :parent_author_signature, default: nil
+          property :author_signature, default: nil
+
+          # Generates XML and updates signatures
+          def to_xml
+            xml = entity_xml
+            hash = to_h
+            Relayable.update_signatures!(hash)
+
+            xml.at_xpath("author_signature").content = hash[:author_signature]
+            xml.at_xpath("parent_author_signature").content = hash[:parent_author_signature]
+            xml
+          end
         end
       end
+
+      class SignatureVerificationFailed < ArgumentError
+      end
+
+      def self.verify_signatures(data)
+        pkey = DiasporaFederation.callbacks.trigger(:fetch_public_key_by_id, data[:diaspora_id])
+        raise SignatureVerificationFailed, "failed to fetch public key for #{data[:diaspora_id]}" if pkey.nil?
+        raise SignatureVerificationFailed, "wrong author_signature" unless Signing.verify_signature(
+          data, data[:author_signature], pkey
+        )
+
+        unless DiasporaFederation.callbacks.trigger(:post_author_is_local?, data[:parent_guid])
+          # this happens only on downstream federation
+          pkey = DiasporaFederation.callbacks.trigger(:fetch_public_key_by_post_guid, data[:parent_guid])
+          raise SignatureVerificationFailed,
+                "failed to fetch public key for parent of #{data[:parent_guid]}" if pkey.nil?
+          raise SignatureVerificationFailed, "wrong parent_author_signature" unless Signing.verify_signature(
+            data, data[:parent_author_signature], pkey
+          )
+        end
+      end
+
+      def self.update_signatures!(data)
+        if data[:author_signature].nil?
```

Comment:
```has_key?``` is not exactly equivalent to that, because a key can be
set while it can still be nil, so it's shorter this way, it combines
both checks in one.